# Яндекс.Музыка

Сравнение Москвы и Петербурга окружено мифами. Например:
 * Москва — мегаполис, подчинённый жёсткому ритму рабочей недели;
 * Петербург — культурная столица, со своими вкусами.

На данных Яндекс.Музыки вы сравните поведение пользователей двух столиц.

**Цель исследования** — проверьте три гипотезы:
1. Активность пользователей зависит от дня недели. Причём в Москве и Петербурге это проявляется по-разному.
2. В понедельник утром в Москве преобладают одни жанры, а в Петербурге — другие. Так же и вечером пятницы преобладают разные жанры — в зависимости от города. 
3. Москва и Петербург предпочитают разные жанры музыки. В Москве чаще слушают поп-музыку, в Петербурге — русский рэп.

**Ход исследования**

Данные о поведении пользователей вы получите из файла `yandex_music_project.csv`. О качестве данных ничего не известно. Поэтому перед проверкой гипотез понадобится обзор данных. 

Вы проверите данные на ошибки и оцените их влияние на исследование. Затем, на этапе предобработки вы поищете возможность исправить самые критичные ошибки данных.
 
Таким образом, исследование пройдёт в три этапа:
 1. Обзор данных.
 2. Предобработка данных.
 3. Проверка гипотез.



## Обзор данных

Составьте первое представление о данных Яндекс.Музыки.




Основной инструмент аналитика — `pandas`. Импортируйте эту библиотеку.

In [1]:
# импорт библиотеки pandas
import pandas as pd

Прочитайте файл `yandex_music_project.csv` из папки `/datasets` и сохраните его в переменной `df`:

In [2]:
# чтение файла с данными и сохранение в df
df = pd.read_csv('/datasets/yandex_music_project.csv')

Выведите на экран первые десять строк таблицы:

In [3]:
# получение первых 10 строк таблицы df
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                    Преданная         IMPERVTOR  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE     И вновь продолжается бой               NaN  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

             City        time        Day  
0  Saint-Petersburg  20:28:33  Wednesday  
1            Moscow  14:07:09     Friday  
2  Saint-Petersburg  20:58:07  Wednesday  
3  Saint-Petersburg  08:37:09     Monday  
4            Moscow  08:34:34     Monday  
5  Saint-Petersburg  13:09:41     Friday  
6            Moscow  13:00:07  Wednesday  
7            Moscow  20:47:49  Wednesday  
8            Moscow  09:17:40     Friday  
9  Saint-Petersburg  21:20:49  Wednesday

In [4]:
# Комментарий ревьюера
df.tail(5)

userID                              Track         artist       genre  \
65074  729CBB09                            My Name         McLean         rnb   
65075  D08D4A55  Maybe One Day (feat. Black Spade)    Blu & Exile         hip   
65076  C5E3A0D5                          Jalopiina            NaN  industrial   
65077  321D0506                      Freight Train  Chas McDevitt        rock   
65078  3A64EF84          Tell Me Sweet Little Lies   Monica Lopez     country   

                 City        time        Day  
65074            Moscow  13:32:28  Wednesday  
65075  Saint-Petersburg  10:00:00     Monday  
65076            Moscow  20:09:26     Friday  
65077            Moscow  21:43:59     Friday  
65078            Moscow  21:59:46     Friday

Одной командой получить общую информацию о таблице:

In [5]:
# получение общей информации о данных в таблице df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
  userID    65079 non-null object
Track       63848 non-null object
artist      57876 non-null object
genre       63881 non-null object
  City      65079 non-null object
time        65079 non-null object
Day         65079 non-null object
dtypes: object(7)
memory usage: 3.5+ MB


Итак, в таблице семь столбцов. Тип данных во всех столбцах — `object`.

Согласно документации к данным:
* `userID` — идентификатор пользователя;
* `Track` — название трека;  
* `artist` — имя исполнителя;
* `genre` — название жанра;
* `City` — город пользователя;
* `time` — время начала прослушивания;
* `Day` — день недели.

В названиях колонок видны три нарушения стиля:
1. Строчные буквы сочетаются с прописными.
2. Встречаются пробелы.
3. Не разделены подчеркиваниями части названия (пример: 'userID' вместо 'user_id')


Количество значений в столбцах различается. Значит, в данных есть пропущенные значения.


**Выводы**

В каждой строке таблицы — данные о прослушанном треке. Часть колонок описывает саму композицию: название, исполнителя и жанр. Остальные данные рассказывают о пользователе: из какого он города, когда он слушал музыку. 

Предварительно можно утверждать, что, данных достаточно для проверки гипотез. Но встречаются пропуски в данных, а в названиях колонок — расхождения с хорошим стилем.

Чтобы двигаться дальше, нужно устранить проблемы в данных.

## Предобработка данных
Исправьте стиль в заголовках столбцов, исключите пропуски. Затем проверьте данные на дубликаты.

### Стиль заголовков
Выведите на экран названия столбцов:

In [6]:
# перечень названий столбцов таблицы df
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

Приведите названия в соответствие с хорошим стилем:
* несколько слов в названии запишите в «змеином_регистре»,
* все символы сделайте строчными,
* устраните пробелы.

Для этого переименуйте колонки так:
* `'  userID'` → `'user_id'`;
* `'Track'` → `'track'`;
* `'  City  '` → `'city'`;
* `'Day'` → `'day'`.

In [7]:
# переименование столбцов
df=df.rename(columns={'  userID':'user_id','Track':'track','  City  ':'city','Day':'day'})

Проверьте результат. Для этого ещё раз выведите на экран названия столбцов:

In [8]:
# проверка результатов - перечень названий столбцов
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

### Пропуски значений
Сначала посчитайте, сколько в таблице пропущенных значений. Для этого достаточно двух методов `pandas`:

In [9]:
# подсчёт пропусков
df.isna().sum()

user_id       0
track      1231
artist     7203
genre      1198
city          0
time          0
day           0
dtype: int64

Не все пропущенные значения влияют на исследование. Так в `track` и `artist` пропуски не важны для вашей работы. Достаточно заменить их явными обозначениями.

Но пропуски в `genre` могут помешать сравнению музыкальных вкусов в Москве и Санкт-Петербурге. На практике было бы правильно установить причину пропусков и восстановить данные. Такой возможности нет в учебном проекте. Придётся:
* заполнить и эти пропуски явными обозначениями,
* оценить, насколько они повредят расчётам. 

Замените пропущенные значения в столбцах `track`, `artist` и `genre` на строку `'unknown'`. Для этого создайте список `columns_to_replace`, переберите его элементы циклом `for` и для каждого столбца выполните замену пропущенных значений:

In [10]:
# перебор названий столбцов в цикле и замена пропущенных значений на 'unknown'

columns_to_replace = ['track','artist','genre']

for _ in columns_to_replace:
    df[_] = df[_].fillna('unknown')
    

Убедитесь, что в таблице не осталось пропусков. Для этого ещё раз посчитайте пропущенные значения.

In [11]:
# подсчёт пропусков
df.isna().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

### Дубликаты
Посчитайте явные дубликаты в таблице одной командой:

In [12]:
# подсчёт явных дубликатов
df.duplicated().sum()

3826

Вызовите специальный метод `pandas`, чтобы удалить явные дубликаты:

In [13]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых)
df = df.drop_duplicates().reset_index()

Ещё раз посчитайте явные дубликаты в таблице — убедитесь, что полностью от них избавились:

In [14]:
# проверка на отсутствие дубликатов
df.duplicated().sum()

0

Теперь избавьтесь от неявных дубликатов в колонке `genre`. Например, название одного и того же жанра может быть записано немного по-разному. Такие ошибки тоже повлияют на результат исследования.

Выведите на экран список уникальных названий жанров, отсортированный в алфавитном порядке. Для этого:
* извлеките нужный столбец датафрейма, 
* примените к нему метод сортировки,
* для отсортированного столбца вызовите метод, который вернёт уникальные значения из столбца.

In [15]:
# Просмотр уникальных названий жанров
genre_column = df['genre'].sort_values()
genre_column = genre_column.unique()
genre_column

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'alternativepunk', 'ambient', 'americana',
       'animated', 'anime', 'arabesk', 'arabic', 'arena',
       'argentinetango', 'art', 'audiobook', 'author', 'avantgarde',
       'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass',
       'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks',
       'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean',
       'caucasian', 'celtic', 'chamber', 'chanson', 'children', 'chill',
       'chinese', 'choral', 'christian', 'christmas', 'classical',
       'classicmetal', 'club', 'colombian', 'comedy', 'conjazz',
       'contemporary', 'country', 'cuban', 'dance', 'dancehall',
       'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr',
       'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo',
       'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic',
       'electropop', 'emo', 'entehno', '

Просмотрите список и найдите неявные дубликаты названия `hiphop`. Это могут быть названия с ошибками или альтернативные названия того же жанра.

Вы увидите следующие неявные дубликаты:
* *hip*,
* *hop*,
* *hip-hop*.

Чтобы очистить от них таблицу, напишите функцию `replace_wrong_genres()` с двумя параметрами: 
* `wrong_genres` — список дубликатов,
* `correct_genre` — строка с правильным значением.

Функция должна исправить колонку `genre` в таблице `df`: заменить каждое значение из списка `wrong_genres` на значение из `correct_genre`.

In [16]:
# Функция для замены неявных дубликатов
def replace_wrong_genres(wrong_genres, correct_genre):
    for _ in wrong_genres:
        df['genre']=df['genre'].replace(_,correct_genre)

Вызовите `replace_wrong_genres()` и передайте ей такие аргументы, чтобы она устранила неявные дубликаты: вместо `hip`, `hop` и `hip-hop` в таблице должно быть значение `hiphop`:

In [17]:
# Устранение неявных дубликатов
replace_wrong_genres(['hip', 'hop', 'hip-hop'],'hiphop')

Проверьте, что заменили неправильные названия:

*   hip
*   hop
*   hip-hop

Выведите отсортированный список уникальных значений столбца `genre`:

In [18]:
# Проверка на неявные дубликаты
genre_column = df['genre'].sort_values().unique()
print(genre_column)

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'alternativepunk' 'ambient' 'americana' 'animated' 'anime' 'arabesk'
 'arabic' 'arena' 'argentinetango' 'art' 'audiobook' 'author' 'avantgarde'
 'axé' 'baile' 'balkan' 'beats' 'bigroom' 'black' 'bluegrass' 'blues'
 'bollywood' 'bossa' 'brazilian' 'breakbeat' 'breaks' 'broadway'
 'cantautori' 'cantopop' 'canzone' 'caribbean' 'caucasian' 'celtic'
 'chamber' 'chanson' 'children' 'chill' 'chinese' 'choral' 'christian'
 'christmas' 'classical' 'classicmetal' 'club' 'colombian' 'comedy'
 'conjazz' 'contemporary' 'country' 'cuban' 'dance' 'dancehall' 'dancepop'
 'dark' 'death' 'deep' 'deutschrock' 'deutschspr' 'dirty' 'disco' 'dnb'
 'documentary' 'downbeat' 'downtempo' 'drum' 'dub' 'dubstep' 'eastern'
 'easy' 'electronic' 'electropop' 'emo' 'entehno' 'epicmetal' 'estrada'
 'ethnic' 'eurofolk' 'european' 'experimental' 'extrememetal' 'fado'
 'fairytail' 'film' 'fitness' 'flamenco' 'folk' 'folklore' 'folkmetal'
 'folkrock' 

**Выводы**

Предобработка обнаружила три проблемы в данных:

- нарушения в стиле заголовков,
- пропущенные значения,
- дубликаты — явные и неявные.

Вы исправили заголовки, чтобы упростить работу с таблицей. Без дубликатов исследование станет более точным.

Пропущенные значения вы заменили на `'unknown'`. Ещё предстоит увидеть, не повредят ли исследованию пропуски в колонке `genre`.

Теперь можно перейти к проверке гипотез. 

## Проверка гипотез

### Сравнение поведения пользователей двух столиц

Первая гипотеза утверждает, что пользователи по-разному слушают музыку в Москве и Санкт-Петербурге. Проверьте это предположение по данным о трёх днях недели — понедельнике, среде и пятнице. Для этого:

* Разделите пользователей Москвы и Санкт-Петербурга
* Сравните, сколько треков послушала каждая группа пользователей в понедельник, среду и пятницу.


Для тренировки сначала выполните каждый из расчётов по отдельности. 

Оцените активность пользователей в каждом городе. Сгруппируйте данные по городу и посчитайте прослушивания в каждой группе.



In [19]:
# Подсчёт прослушиваний в каждом городе
print('В Москве прослушиваний -', df[df['city']=='Moscow']['city'].count(), 'доля unknown', df[(df['city']=='Moscow') & (df['genre']=='unknown')]['city'].count()/df[df['city']=='Moscow']['city'].count()*100,'%')
print('В Петербурге прослушиваний -', df[df['city']=='Saint-Petersburg']['city'].count(), 'доля unknown', df[(df['city']=='Saint-Petersburg') & (df['genre']=='unknown')]['city'].count()/df[df['city']=='Saint-Petersburg']['city'].count()*100,'%')

В Москве прослушиваний - 42741 доля unknown 1.9863830981961115 %
В Петербурге прослушиваний - 18512 доля unknown 1.5017286084701815 %


В Москве прослушиваний больше, чем в Петербурге. Из этого не следует, что московские пользователи чаще слушают музыку. Просто самих пользователей в Москве больше.

Теперь сгруппируйте данные по дню недели и подсчитайте прослушивания в понедельник, среду и пятницу. Учтите, что в данных есть информация только о прослушиваниях только за эти дни.


In [20]:
# Подсчёт прослушиваний в каждый из трёх дней
df.groupby('day')['city'].count()

day
Friday       21840
Monday       21354
Wednesday    18059
Name: city, dtype: int64

В среднем пользователи из двух городов менее активны по средам. Но картина может измениться, если рассмотреть каждый город в отдельности.

Вы видели, как работает группировка по городу и по дням недели. Теперь напишите функцию, которая объединит два эти расчёта.

Создайте функцию `number_tracks()`, которая посчитает прослушивания для заданного дня и города. Ей понадобятся два параметра:
* день недели,
* название города.

В функции сохраните в переменную строки исходной таблицы, у которых значение:
  * в колонке `day` равно параметру `day`,
  * в колонке `city` равно параметру `city`.

Для этого примените последовательную фильтрацию с логической индексацией.

Затем посчитайте значения в столбце `user_id` получившейся таблицы. Результат сохраните в новую переменную. Верните эту переменную из функции.

In [21]:
# <создание функции number_tracks()>
# Объявляется функция с двумя параметрами: day, city.
def number_tracks(day,city):
    track_list = df[(df['city']==city)&(df['day']==day)]
# В переменной track_list сохраняются те строки таблицы df, для которых 
# значение в столбце 'day' равно параметру day и одновременно значение
# в столбце 'city' равно параметру city (используйте последовательную фильтрацию
# с помощью логической индексации).

# В переменной track_list_count сохраняется число значений столбца 'user_id',
# рассчитанное методом count() для таблицы track_list.
# Функция возвращает число - значение track_list_count.
    track_list_count = track_list['user_id'].count()
    return track_list_count

# Функция для подсчёта прослушиваний для конкретного города и дня.
# С помощью последовательной фильтрации с логической индексацией она 
# сначала получит из исходной таблицы строки с нужным днём,
# затем из результата отфильтрует строки с нужным городом,
# методом count() посчитает количество значений в колонке user_id. 
# Это количество функция вернёт в качестве результата
peterburg_data = ['Saint-Petersburg']
moscow_data = ['Moscow']


Вызовите `number_tracks()` шесть раз, меняя значение параметров — так, чтобы получить данные для каждого города в каждый из трёх дней.

In [22]:
# количество прослушиваний в Москве по понедельникам
moscow_data.append(number_tracks('Monday',"Moscow"))

In [23]:
# количество прослушиваний в Санкт-Петербурге по понедельникамnumber_tracks('Monday',"Moscow")
peterburg_data.append(number_tracks('Monday','Saint-Petersburg'))

In [24]:
# количество прослушиваний в Москве по средам
moscow_data.append(number_tracks('Wednesday',"Moscow"))

In [25]:
# количество прослушиваний в Санкт-Петербурге по средам
peterburg_data.append(number_tracks('Wednesday','Saint-Petersburg'))

In [26]:
# количество прослушиваний в Москве по пятницам
moscow_data.append(number_tracks('Friday',"Moscow"))

In [27]:
# количество прослушиваний в Санкт-Петербурге по пятницам
peterburg_data.append(number_tracks('Friday','Saint-Petersburg'))

Создайте c помощью конструктора `pd.DataFrame` таблицу, где
* названия колонок — `['city', 'monday', 'wednesday', 'friday']`;
* данные — результаты, которые вы получили с помощью `number_tracks`.

In [28]:
# Таблица с результатами
cities_in_days = [moscow_data,peterburg_data]
columns = ['city', 'monday', 'wednesday', 'friday']
listenings_by_days_cities=pd.DataFrame(columns = columns,data = cities_in_days)
listenings_by_days_cities

city  monday  wednesday  friday
0            Moscow   15740      11056   15945
1  Saint-Petersburg    5614       7003    5895

**Выводы**

Данные показывают разницу поведения пользователей:

- В Москве пик прослушиваний приходится на понедельник и пятницу, а в среду заметен спад.
- В Петербурге, наоборот, больше слушают музыку по средам. Активность в понедельник и пятницу здесь почти в равной мере уступает среде.

Значит, данные говорят в пользу первой гипотезы.

### Музыка в начале и в конце недели

Согласно второй гипотезе, утром в понедельник в Москве преобладают одни жанры, а в Петербурге — другие. Так же и вечером пятницы преобладают разные жанры — в зависимости от города.

Сохраните таблицы с данными в две переменные:
* по Москве — в `moscow_general`;
* по Санкт-Петербургу — в `spb_general`.

In [29]:
# получение таблицы moscow_general из тех строк таблицы df, 
# для которых значение в столбце 'city' равно 'Moscow'
moscow_general = df[df['city']=='Moscow']
moscow_general

index   user_id                          track  \
1          1  55204538    Delayed Because of Accident   
4          4  E2DC1FAE                    Soul People   
6          6  4CB90AA5                           True   
7          7  F03E1C1F               Feeling This Way   
8          8  8FA1D3BE       И вновь продолжается бой   
...      ...       ...                            ...   
61247  65073  83A474E7  I Worship Only What You Bleed   
61248  65074  729CBB09                        My Name   
61250  65076  C5E3A0D5                      Jalopiina   
61251  65077  321D0506                  Freight Train   
61252  65078  3A64EF84      Tell Me Sweet Little Lies   

                        artist         genre    city      time        day  
1             Andreas Rönnberg          rock  Moscow  14:07:09     Friday  
4                   Space Echo         dance  Moscow  08:34:34     Monday  
6                 Roman Messer         dance  Moscow  13:00:07  Wednesday  
7              Polina Griffith         dance  Moscow  20:47:49  Wednesday  
8                      unknown        ruspop  Moscow  09:17:40     Friday  
...                        ...           ...     ...       ...        ...  
61247  The Black Dahlia Murder  extrememetal  Moscow  21:07:12     Monday  
61248                   McLean           rnb  Moscow  13:32:28  Wednesday  
61250                  unknown    industrial  Moscow  20:09:26     Friday  
61251            Chas McDevitt          rock  Moscow  21:43:59     Friday  
61252             Monica Lopez       country  Moscow  21:59:46     Friday  

[42741 rows x 8 columns]

In [30]:
# получение таблицы spb_general из тех строк таблицы df,
# для которых значение в столбце 'city' равно 'Saint-Petersburg'
spb_general = df[df['city']=='Saint-Petersburg']
spb_general

index   user_id                              track              artist  \
0          0  FFB692EC                  Kamigata To Boots    The Mass Missile   
2          2    20EC38                  Funiculì funiculà         Mario Lanza   
3          3  A3DD03C9              Dragons in the Sunset          Fire + Ice   
5          5  842029A1                          Преданная           IMPERVTOR   
9          9  E772D5C0                          Pessimist             unknown   
...      ...       ...                                ...                 ...   
61239  65063  D94F810B        Theme from the Walking Dead  Proyecto Halloween   
61240  65064  BC8EC5CF       Red Lips: Gta (Rover Rework)               Rover   
61241  65065  29E04611                       Bre Petrunko       Perunika Trio   
61242  65066  1B91C621             (Hello) Cloud Mountain     sleepmakeswaves   
61249  65075  D08D4A55  Maybe One Day (feat. Black Spade)         Blu & Exile   

            genre              city      time        day  
0            rock  Saint-Petersburg  20:28:33  Wednesday  
2             pop  Saint-Petersburg  20:58:07  Wednesday  
3            folk  Saint-Petersburg  08:37:09     Monday  
5          rusrap  Saint-Petersburg  13:09:41     Friday  
9           dance  Saint-Petersburg  21:20:49  Wednesday  
...           ...               ...       ...        ...  
61239        film  Saint-Petersburg  21:14:40     Monday  
61240  electronic  Saint-Petersburg  21:06:50     Monday  
61241       world  Saint-Petersburg  13:56:00     Monday  
61242    postrock  Saint-Petersburg  09:22:13     Monday  
61249      hiphop  Saint-Petersburg  10:00:00     Monday  

[18512 rows x 8 columns]

Создайте функцию `genre_weekday()` с четырьмя параметрами:
* таблица (датафрейм) с данными,
* день недели,
* начальная временная метка в формате 'hh:mm', 
* последняя временная метка в формате 'hh:mm'.

Функция должна вернуть информацию о топ-10 жанров тех треков, которые прослушивали в указанный день, в промежутке между двумя отметками времени.

In [31]:
# Объявление функции genre_weekday() с параметрами table, day, time1, time2,
def genre_weekday(table, day, time1, time2):
# которая возвращает информацию о самых популярных жанрах в указанный день в
# заданное время:
# 1) в переменную genre_df сохраняются те строки переданного датафрейма table, для
#    которых одновременно:
#    - значение в столбце day равно значению аргумента day
#    - значение в столбце time больше значения аргумента time1
#    - значение в столбце time меньше значения аргумента time2

    
#    Используйте последовательную фильтрацию с помощью логической индексации.
# последовательная это вроде вот:
    table = table[table['day']==day]
    table = table[table['time']>time1]
    table = table[table['time']<time2]
# 2) сгруппировать датафрейм genre_df по столбцу genre, взять один из его
#    столбцов и посчитать методом count() количество записей для каждого из
#    присутствующих жанров, получившийся Series записать в переменную
#    genre_df_count
    genre_df_count = table.groupby('genre')['day'].count()
# 3) отсортировать genre_df_count по убыванию встречаемости и сохранить
#    в переменную genre_df_sorted
    genre_df_sorted = genre_df_count.sort_values(ascending = False)
# 4) вернуть Series из 10 первых значений genre_df_sorted, это будут топ-10
#    популярных жанров (в указанный день, в заданное время)
    return genre_df_sorted[:10]


Cравните результаты функции `genre_weekday()` для Москвы и Санкт-Петербурга в понедельник утром (с 7:00 до 11:00) и в пятницу вечером (с 17:00 до 23:00):

In [32]:
# вызов функции для утра понедельника в Москве (вместо df — таблица moscow_general)
# объекты, хранящие время, являются строками и сравниваются как строки
# пример вызова: genre_weekday(moscow_general, 'Monday', '07:00', '11:00')
genre_and_count=genre_weekday(moscow_general, 'Monday', '07:00', '11:00')
#Создаем таблицу сравнения Понедельник-Птяница/Москва-Петербург - 'table_day_city_count'
table_day_city_count = pd.DataFrame()
#Создаем/обуляем и наполняем таблицу с результатам подсчета "за день по одному городу" - "one_day_city_count"
one_day_city_count = pd.DataFrame()
one_day_city_count['Жанр']=genre_and_count.index
one_day_city_count['Москва-Понедельник']=genre_and_count.values
#Присоединяем результаты за день к общей таблице сравнения 
table_day_city_count = pd.concat([table_day_city_count, one_day_city_count],sort=False).reset_index(drop = True)
# Выводим результаты за день.
genre_and_count = pd.DataFrame(genre_and_count)
genre_and_count

day
genre           
pop          781
dance        549
electronic   480
rock         474
hiphop       286
ruspop       186
world        181
rusrap       175
alternative  164
unknown      161

In [33]:
# вызов функции для утра понедельника в Петербурге (вместо df — таблица spb_general)
genre_and_count=genre_weekday(spb_general, 'Monday', '07:00', '11:00')
#Создаем/обуляем и наполняем таблицу с результатам подсчета "за день по одному городу" - "one_day_city_count"
one_day_city_count = pd.DataFrame()
one_day_city_count['Жанр']=genre_and_count.index
one_day_city_count['Петербург-Понедельник']=genre_and_count.values
#Присоединяем результаты за день к общей таблице сравнения 
table_day_city_count = pd.concat([table_day_city_count, one_day_city_count],sort=False).reset_index(drop = True)
# Выводим результаты за день.
genre_and_count = pd.DataFrame(genre_and_count)
genre_and_count

day
genre           
pop          218
dance        182
rock         162
electronic   147
hiphop        80
ruspop        64
alternative   58
rusrap        55
jazz          44
classical     40

In [34]:
# вызов функции для вечера пятницы в Москве
genre_and_count=genre_weekday(moscow_general, 'Friday', '17:00', '23:00') 
#проверил, с 17:00 до 20:00 значений в исходных данных ни по Москве ни по Санкт-Петербургу нет.

#Создаем/обуляем и наполняем таблицу с результатам подсчета "за день по одному городу" - "one_day_city_count"
one_day_city_count = pd.DataFrame()
one_day_city_count['Жанр']=genre_and_count.index
one_day_city_count['Москва-Пятница']=genre_and_count.values
#Присоединяем результаты за день к общей таблице сравнения 
table_day_city_count = pd.concat([table_day_city_count, one_day_city_count],sort=False).reset_index(drop = True)
# Выводим результаты за день.
genre_and_count = pd.DataFrame(genre_and_count)
genre_and_count

day
genre           
pop          713
rock         517
dance        495
electronic   482
hiphop       273
world        208
ruspop       170
alternative  163
classical    163
rusrap       142

In [35]:
# вызов функции для вечера пятницы в Петербурге
genre_and_count=genre_weekday(spb_general, 'Friday', '17:00', '23:00')
#Создаем/обуляем и наполняем таблицу с результатам подсчета "за день по одному городу" - "one_day_city_count"
one_day_city_count = pd.DataFrame()
one_day_city_count['Жанр']=genre_and_count.index
one_day_city_count['Петербург-Пятница']=genre_and_count.values
#Присоединяем результаты за день к общей таблице сравнения 
table_day_city_count = pd.concat([table_day_city_count, one_day_city_count],sort=False).reset_index(drop = True)
# Выводим результаты за день.
genre_and_count = pd.DataFrame(genre_and_count)
display(genre_and_count)
#Выводим общую таблицу
table_day_city_count.groupby('Жанр').sum()

day
genre           
pop          256
rock         216
electronic   216
dance        210
hiphop        97
alternative   63
jazz          61
classical     60
rusrap        59
world         54

Москва-Понедельник  Петербург-Понедельник  Москва-Пятница  \
Жанр                                                                     
alternative               164.0                   58.0           163.0   
classical                   0.0                   40.0           163.0   
dance                     549.0                  182.0           495.0   
electronic                480.0                  147.0           482.0   
hiphop                    286.0                   80.0           273.0   
jazz                        0.0                   44.0             0.0   
pop                       781.0                  218.0           713.0   
rock                      474.0                  162.0           517.0   
ruspop                    186.0                   64.0           170.0   
rusrap                    175.0                   55.0           142.0   
unknown                   161.0                    0.0             0.0   
world                     181.0                    0.0           208.0   

             Петербург-Пятница  
Жанр                            
alternative               63.0  
classical                 60.0  
dance                    210.0  
electronic               216.0  
hiphop                    97.0  
jazz                      61.0  
pop                      256.0  
rock                     216.0  
ruspop                     0.0  
rusrap                    59.0  
unknown                    0.0  
world                     54.0

**Выводы**

Если сравнить топ-10 жанров в понедельник утром, можно сделать такие выводы:

1. В Москве и Петербурге слушают похожую музыку. Единственное отличие — только в московский рейтинг вошел жанр “world”, и только в петербургский рейтинг вошли жанры — джаз, классика.

2. В Москве пропущенных значений оказалось так много, что значение `'unknown'` заняло десятое место среди самых популярных жанров. Значит, пропущенные значения занимают существенную долю в данных и угрожают достоверности исследования.

Вечер пятницы не меняет эту картину. Некоторые жанры поднимаются немного выше, другие спускаются, но в целом топ-10 остаётся тем же самым.

Таким образом, вторая гипотеза подтвердилась лишь частично:
* Пользователи слушают похожую музыку в начале недели и в конце.
* Разница между Москвой и Петербургом не слишком выражена. В Москве чаще слушают русскую популярную музыку, в Петербурге — джаз.

Однако пропуски в данных ставят под сомнение этот результат. В Москве их так много, что рейтинг топ-10 мог бы выглядеть иначе, если бы не утерянные  данные о жанрах.

### Жанровые предпочтения в Москве и Петербурге

Гипотеза: Петербург — столица рэпа, музыку этого жанра там слушают чаще, чем в Москве.  А Москва — город контрастов, в котором, тем не менее, преобладает поп-музыка.

Сгруппируйте таблицу `moscow_general` по жанру и посчитайте прослушивания треков каждого жанра методом `count()`. Затем отсортируйте результат в порядке убывания и сохраните его в таблице `moscow_genres`.

In [36]:
# одной строкой: группировка таблицы moscow_general по столбцу 'genre', 
# подсчёт числа значений 'genre' в этой группировке методом count(), 
# сортировка получившегося Series в порядке убывания и сохранение в moscow_genres
moscow_genres = moscow_general.groupby('genre')['genre'].count().sort_values(ascending=False)

# создаем таблицу для сравнения количества/интенсивности прослушивания по городам

one_city_count = pd.DataFrame()
one_city_count['Жанр МСК']=moscow_genres.head(10).index
one_city_count['Кол-во МСК']=moscow_genres.head(10).values
#считаем долю прослушиваний всего от количества прослушиваний по городу:
one_city_count['Доля по МСК']= list(map(lambda x: '%2.2f%%' % (x / moscow_genres.sum()*100),one_city_count['Кол-во МСК']))

Выведите на экран первые десять строк `moscow_genres`:

In [37]:
# просмотр первых 10 строк moscow_genres
display(moscow_genres.head(10))
# немного контрольной статистики
display('Доля первых трех по популярности жанров - ' + str(moscow_genres.head(3).sum()/moscow_genres.sum()*100)+' %')

genre
pop            5892
dance          4435
rock           3965
electronic     3786
hiphop         2096
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: genre, dtype: int64

'Доля первых трех по популярности жанров - 33.43861865655928 %'

Теперь повторите то же и для Петербурга.

Сгруппируйте таблицу `spb_general` по жанру. Посчитайте прослушивания треков каждого жанра. Результат отсортируйте в порядке убывания и сохраните в таблице `spb_genres`:


In [38]:
# одной строкой: группировка таблицы spb_general по столбцу 'genre', 
# подсчёт числа значений 'genre' в этой группировке методом count(), 
# сортировка получившегося Series в порядке убывания и сохранение в spb_genres
spb_genres = spb_general.groupby('genre')['genre'].count().sort_values(ascending=False)

#Продолжаем наполнять таблицу для сравнения статистики по городам
one_city_count['Жанр СПб']=spb_genres.head(10).index
one_city_count['Кол-во СПб']=spb_genres.head(10).values
#считаем долю прослушиваний всего от количества прослушиваний по городу:
one_city_count['Доля по СПб']= list(map(lambda x: '%2.2f%%' % (x / spb_genres.sum()*100),one_city_count['Кол-во СПб']))




Выведите на экран первые десять строк `spb_genres`:

In [39]:
# просмотр первых 10 строк spb_genres
display(spb_genres.head(10))
display('Доля первых трех по популярности жанров - ' + str(spb_genres.head(3).sum()/spb_genres.sum()*100)+' %')

genre
pop            2431
dance          1932
rock           1879
electronic     1736
hiphop          960
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: genre, dtype: int64

'Доля первых трех по популярности жанров - 33.71866897147796 %'

In [40]:
# выводим общую статистику по городам.
one_city_count

Жанр МСК  Кол-во МСК Доля по МСК     Жанр СПб  Кол-во СПб Доля по СПб
0          pop        5892      13.79%          pop        2431      13.13%
1        dance        4435      10.38%        dance        1932      10.44%
2         rock        3965       9.28%         rock        1879      10.15%
3   electronic        3786       8.86%   electronic        1736       9.38%
4       hiphop        2096       4.90%       hiphop         960       5.19%
5    classical        1616       3.78%  alternative         649       3.51%
6        world        1432       3.35%    classical         646       3.49%
7  alternative        1379       3.23%       rusrap         564       3.05%
8       ruspop        1372       3.21%       ruspop         538       2.91%
9       rusrap        1161       2.72%        world         515       2.78%

**Выводы**

Гипотеза частично подтвердилась:
* Поп-музыка, танцевальная ('dance') и рок — самые популярные жанры как в Москве, так и в Петербурге (33% всех прослушиваний). Более того, в топ-10 жанров встречается близкий жанр — русская популярная музыка.
* Вопреки ожиданиям, рэп одинаково популярен в Москве и Петербурге. 


## Итоги исследования

Вы проверили три гипотезы и установили:

1. День недели по-разному влияет на активность пользователей в Москве и Петербурге. 

Первая гипотеза полностью подтвердилась.

2. Музыкальные предпочтения не сильно меняются в течение недели — будь то Москва или Петербург. Небольшие различия заметны в начале недели, по понедельникам:
* в Москве слушают музыку жанра “world” и "classic" чаще чем в Петербурге
* в Петербурге чаще слушают "alternative" и русских исполнителей.

Таким образом, вторая гипотеза подтвердилась лишь отчасти. Этот результат мог оказаться иным, если бы не пропуски в данных.

3. Во вкусах пользователей Москвы и Петербурга больше общего чем различий. Вопреки ожиданиям, предпочтения жанров в Петербурге напоминают московские.

Третья гипотеза не подтвердилась. Если различия в предпочтениях и существуют, на основной массе пользователей они незаметны.

**На практике исследования содержат проверки статистических гипотез.**
Из данных одного сервиса не всегда можно сделать вывод о всех жителях города.
Проверки статистических гипотез покажут, насколько они достоверны, исходя из имеющихся данных. 
С методами проверок гипотез вы ещё познакомитесь в следующих темах.